### ISSUE TITLE
User has no identifier

### ISSUE THREAD

**Issue by jeremygachet** [floriansemm/SolrBundle#97]
Hi, 
I've got another issue with the bundle. 
I use fosuserbundle to manage my users and i want to index a huge user list (250 000 users)

But the bundle does not find the Identifier annotation on it. 

Here is my declaration

namespace UserBundle\Entity;

use Doctrine\ORM\Mapping as ORM;
use FOS\UserBundle\Entity\User as BaseUser;
use FOS\UserBundle\Model\GroupInterface;
use Symfony\Component\Validator\Constraints as Assert;
use Symfony\Component\Security\Acl\Domain\UserSecurityIdentity;
use FS\SolrBundle\Doctrine\Annotation as Solr;

/**
 * 
 * @Solr\Document(repository="UserBundle\Entity\User")
 * 
 * @ORM\Table(name="fos_user", indexes={@ORM\Index(name="search_idx", columns={"licenseId", "firstname", "lastname"})})
 * @ORM\Entity(repositoryClass="UserBundle\Repository\UserRepository")
 */
class User extends BaseUser
{

    /**
     * @var int
     * @Solr\Id()
     * @ORM\Column(name="id", type="integer")
     * @ORM\Id()
     * @ORM\GeneratedValue(strategy="AUTO")
     * 
     */
    protected $id;

......

Everything works well everywhere, but not here. 

Plus, is it possible to create batchs ? I need to do a findAll on 250 000 users, it simply block my computer. 

Thank you in advance,

**Comment by jeremygachet** [comment]
For the batch update, I modify (just a little bit) your command. 
I just set batches, and I switch from ObjectManager to EntityManager (i work with mysql) to be able to detach entities just after the synch. 
Then i don't have memory leaks, so I can synchronize my 250 000 users with ease (and a few hours)


Here is the command code I use, I wish it will help: 



```
<?php

namespace DummyBundle\Command;

use Doctrine\Common\Persistence\AbstractManagerRegistry;
use FS\SolrBundle\Console\ConsoleErrorListOutput;
use Symfony\Bundle\FrameworkBundle\Command\ContainerAwareCommand;
use Symfony\Component\Console\Input\InputArgument;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Output\OutputInterface;

class SolrSynchCommand extends ContainerAwareCommand
{

    private $total = 0;

    protected function configure()
    {
        $this->setName('solr:synchronize')
                ->addArgument('entity', InputArgument::REQUIRED, 'The entity you want to index')
                ->addOption(
                        'source', null, InputArgument::OPTIONAL, 'specify a source from where to load entities [relational, mongodb]', 'relational'
                )
                ->setDescription('Index all entities');
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {
        $entity = $input->getArgument('entity');
        $source = $input->getOption('source');
  //      $objectManager = $this->getObjectManager($source);
        
        $entityManager = $this->getContainer()->get('doctrine')->getEntityManager();
      //  $objectManager = $entityManager->getRepository($source);


        $batchSize = 1000;

        try {
            $repository =  $entityManager->getRepository($entity);
        } catch (\Exception $e) {
            $output->writeln(sprintf('<error>No repository found for "%s", check your input</error>', $entity));

            return;
        }

        //get total quantity
        $qb = $repository->createQueryBuilder('entity');
        $qb->select('COUNT(entity)');

        $total = $qb->getQuery()->getSingleScalarResult();



        $loopQuantity = $total / $batchSize;

        $output->writeln('<info>' . $total . ' entities to index</info>');
        $output->writeln('<info>' . $loopQuantity . ' batches</info>');



        //loop on batch and index them
        for ($i = 0; $i <= $loopQuantity; $i++) {

            $entities = $repository->findBy(array(), null, $batchSize, ($i * $batchSize));


            if (count($entities) == 0) {
                $output->writeln('<comment>No entities found for indexing</comment>');

                return;
            }

            $solr = $this->getContainer()->get('solr.client');




            $this->synchBatch($entities, $solr, $output,  $entityManager);
        }


        $output->writeln(sprintf('<comment>Overall: %s</comment>', $this->total));
    }

    
    
    /**
     * 
     * @param DoctrineCollection $entities
     * @param type $solr
     * @param type $output
     */
    protected function synchBatch($entities, $solr, $output,  $entityManager)
    {

        foreach ($entities as $entity) {
            try {
                $solr->synchronizeIndex($entity);
                $entityManager->detach($entity);
            } catch (\Exception $e) {

                $output->writeln('<error>' . $e->getMessage() . '<error>');
            }
        }
        $results = $this->getContainer()->get('solr.console.command.results');
        if ($results->hasErrors()) {
            $output->writeln('<info>Batch Synchronization finished with errors!</info>');
        } else {
            $output->writeln('<info>Batch  Synchronization successful</info>');
        }

        $output->writeln('');
        $output->writeln(sprintf('<comment>Synchronized Documents: %s</comment>', $results->getSucceed()));
        $output->writeln(sprintf('<comment>Not Synchronized Documents: %s</comment>', $results->getErrored()));
        $output->writeln('');

        $this->total += $results->getOverall();

        if ($results->hasErrors()) {
            $errorList = new ConsoleErrorListOutput($output, $this->getHelper('table'), $results->getErrors());
            $errorList->render();
        }
    }

    /**
     * @param string $source
     * @throws \InvalidArgumentException if $source is unknown
     * @throws \RuntimeException if no doctrine instance is configured
     * @return AbstractManagerRegistry
     */
    private function getObjectManager($source)
    {
        $objectManager = null;

        if ($source == 'relational') {
            $objectManager = $this->getContainer()->get('doctrine');
        } else {
            if ($source == 'mongodb') {
                $objectManager = $this->getContainer()->get('doctrine_mongodb');
            } else {
                throw new \InvalidArgumentException(sprintf('Unknown source %s', $source));
            }
        }

        return $objectManager;
    }

}

```

Thank you again for your great job

**Comment by floriansemm** [comment]
sorry for the late response: a batch-import would be a useful feature

**Comment by jeremygachet** [comment]
No problem, I had to do small improvement to synchronize my 250 000 users. 
First of all, I had to remove doctrine's memory leaks adding :
```   
    $entityManager->getConnection()->getConfiguration()->setSQLLogger(null);
```
and 
```
        if ($this->getContainer()->has('doctrine.dbal.logger.profiling.default')) {
            $this->getContainer()->get('doctrine.dbal.logger.profiling.default')->enabled = false;
        }
```

It really speeds up the command.

I also needed to put a flag on the entity (because I just can't sync 250 000 users each night., I need to make the difference between the users already sync and the new ones)
And so I need to do a request by entitiy.
I think it could be improved by creating a query by batch.

On little thing that accelerate the queries is to use the --no-debug option when I launch the task.

But now it works. It's a bit slow but it's ok.
Thank you again for your work !
